Relax Inc. makes productivity and project management software that's popular with both individuals and teams. Founded by several former Facebook employees, it's considered a great company to work at.

We are given a user table ( "takehome_users" ) with data on 12,000 users who signed up for the product in the last two years and a usage summary table ( "takehome_user_engagement" ) that has a row for each day that a user logged into the product.

The goal is to identify which factors predict future user adoption. An "adopted user" is a user who has logged into the product on three separate days in at least one seven-day period.

The problem can be seen as a binary classification problem which is to predict whether a user is an "adopted user" (class 1) or not (class 0). However in this notebook I also define a concept by the name "adoption score" which is positively related to user activity. Therefore one can also restate the problem as trying to predict what the "adoption score" of a user is. In such a situation the problem will be a regression instead of classification. 

In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

Open data files and read content into two Pandas dataframes. users_df which contains information about each user and eng_df whcih contains engagement(i.e. log-in) data of all users

In [3]:
with open('takehome_users.csv') as users_file, open('takehome_user_engagement.csv') as eng_file:
    users_df = pd.read_csv(users_file)
    eng_df = pd.read_csv(eng_file)

Set the user ID as the index of the users_df dataframe

In [4]:
users_df.set_index('object_id', inplace=True)

In [5]:
users_df

creation_time               name                       email  \
object_id                                                                       
1          2014-04-22 03:53:30     Clausen August    AugustCClausen@yahoo.com   
2          2013-11-15 03:45:04      Poole Matthew      MatthewPoole@gustr.com   
3          2013-03-19 23:14:52  Bottrill Mitchell  MitchellBottrill@gustr.com   
4          2013-05-21 08:09:28    Clausen Nicklas   NicklasSClausen@yahoo.com   
5          2013-01-17 10:14:20          Raw Grace          GraceRaw@yahoo.com   
...                        ...                ...                         ...   
11996      2013-09-06 06:14:15       Meier Sophia       SophiaMeier@gustr.com   
11997      2013-01-10 18:28:37      Fisher Amelie      AmelieFisher@gmail.com   
11998      2014-04-27 12:45:16        Haynes Jake         JakeHaynes@cuvox.de   
11999      2012-05-31 11:55:59       Faber Annett          mhaerzxp@iuxiw.com   
12000      2014-01-26 08:57:12         Lima Thaís   ThaisMeloLima@hotmail.com   

              creation_source  last_session_creation_time  \
object_id                                                   
1                GUEST_INVITE                1.398139e+09   
2                  ORG_INVITE                1.396238e+09   
3                  ORG_INVITE                1.363735e+09   
4                GUEST_INVITE                1.369210e+09   
5                GUEST_INVITE                1.358850e+09   
...                       ...                         ...   
11996              ORG_INVITE                1.378448e+09   
11997      SIGNUP_GOOGLE_AUTH                1.358275e+09   
11998            GUEST_INVITE                1.398603e+09   
11999       PERSONAL_PROJECTS                1.338638e+09   
12000                  SIGNUP                1.390727e+09   

           opted_in_to_mailing_list  enabled_for_marketing_drip  org_id  \
object_id                                                                 
1                                 1                           0      11   
2                                 0                           0       1   
3                                 0                           0      94   
4                                 0                           0       1   
5                                 0                           0     193   
...                             ...                         ...     ...   
11996                             0                           0      89   
11997                             0                           0     200   
11998                             1                           1      83   
11999                             0                           0       6   
12000                             0                           1       0   

           invited_by_user_id  
object_id                      
1                     10803.0  
2                       316.0  
3                      1525.0  
4                      5151.0  
5                      5240.0  
...                       ...  
11996                  8263.0  
11997                     NaN  
11998                  8074.0  
11999                     NaN  
12000                     NaN  

[12000 rows x 9 columns]

In [6]:
users_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 12000 entries, 1 to 12000
Data columns (total 9 columns):
creation_time                 12000 non-null object
name                          12000 non-null object
email                         12000 non-null object
creation_source               12000 non-null object
last_session_creation_time    8823 non-null float64
opted_in_to_mailing_list      12000 non-null int64
enabled_for_marketing_drip    12000 non-null int64
org_id                        12000 non-null int64
invited_by_user_id            6417 non-null float64
dtypes: float64(2), int64(3), object(4)
memory usage: 937.5+ KB


Looking at the above shows:<br>

1) Columns 'last_session_creation_time' and 'invited_by_user_id' have null values<br>
2) 'creation_source' contains categorical data in the form of Strings<br>
3) 'creation_time' and 'last_session_creation_time' are date/time columns but are not of datetime data type.<br>
4) There are 12000 users (i.e rows) and data 10 columns (one of which was set as the index of the data frame) <br>
5) The 12000 user IDs which correspond to the dataframe index are a range starting from 1 to 12000

In [7]:
eng_df

time_stamp  user_id  visited
0       2014-04-22 03:53:30        1        1
1       2013-11-15 03:45:04        2        1
2       2013-11-29 03:45:04        2        1
3       2013-12-09 03:45:04        2        1
4       2013-12-25 03:45:04        2        1
...                     ...      ...      ...
207912  2013-09-06 06:14:15    11996        1
207913  2013-01-15 18:28:37    11997        1
207914  2014-04-27 12:45:16    11998        1
207915  2012-06-02 11:55:59    11999        1
207916  2014-01-26 08:57:12    12000        1

[207917 rows x 3 columns]

In [8]:
eng_df.describe()

user_id   visited
count  207917.000000  207917.0
mean     5913.314197       1.0
std      3394.941674       0.0
min         1.000000       1.0
25%      3087.000000       1.0
50%      5682.000000       1.0
75%      8944.000000       1.0
max     12000.000000       1.0

In [9]:
eng_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 207917 entries, 0 to 207916
Data columns (total 3 columns):
time_stamp    207917 non-null object
user_id       207917 non-null int64
visited       207917 non-null int64
dtypes: int64(2), object(1)
memory usage: 4.8+ MB


The eng_df shows:

1) The 'visited' column is an int64 with 1 as both the maximum and minimum value. This means that the value of 'visited' is 1 for all entries.This column could even be dropped as it is not providing any additional information.<br>
2) time_stamp is not datetime data type.<br>
3) user_id ranges from 1 to 12000.<br>
4) There are 207917 rows of data.

One approch I am taking is to define an adoption score for each user. When traversing the log-in data for each user chronologically I look at 7-day windows of log-in activity. If in a 7-day window a user has logged-in on 3 separate days then I increment the adoption score of that user. The 7-day windows start from the earliest log-in date of the user and the start date of each window is incremented by one during each scan. 

Because there are 207916 rows of data and I want to look at 207916-7+1 = 207910 seven-day-window intervals I decided to import the data into a database using MySQL Workbench. This import took 1 hour and 50 minutes on my laptop. I used the Table Data Import Wizard to import the data. In retrospect it might have been better had I chosen 'datetime' for the time_stamp column instead of 'text'. The table is stored under test_schema local instance schema. 

<img src="image.png">

In [10]:
from sqlalchemy import create_engine

In [11]:
# create engine
engine = create_engine('mysql+pymysql://root:password@localhost/test_schema')

Let's do some more data exploration in SQL

In [12]:
with engine.connect() as con:
    rs = con.execute("SELECT T.user_id AS uid, CAST(T.time_stamp AS DATE) AS dte, COUNT(*) AS num_visits_per_day_per_user " 
                     "FROM takehome_user_engagement AS T "
                     "GROUP BY uid, dte "
                     "HAVING num_visits_per_day_per_user = 1")

    q_1 = pd.DataFrame(rs.fetchall())
    q_1.columns = rs.keys()
    
    rs = con.execute("SELECT T.user_id AS uid, CAST(T.time_stamp AS DATE) AS dte, COUNT(*) AS num_visits_per_day_per_user " 
                     "FROM takehome_user_engagement AS T "
                     "GROUP BY uid, dte "
                     "HAVING num_visits_per_day_per_user > 1")

    q_2 = pd.DataFrame(rs.fetchall())

q_1

uid         dte  num_visits_per_day_per_user
0           1  2014-04-22                            1
1           2  2013-11-15                            1
2           2  2013-11-29                            1
3           2  2013-12-09                            1
4           2  2013-12-25                            1
...       ...         ...                          ...
207912  11996  2013-09-06                            1
207913  11997  2013-01-15                            1
207914  11998  2014-04-27                            1
207915  11999  2012-06-02                            1
207916  12000  2014-01-26                            1

[207917 rows x 3 columns]

In [13]:
q_2

Empty DataFrame
Columns: []
Index: []

The fact that the q_2 query returned an empty result set shows that for each user on a particular day there is at most a single log-in recorded. Now let us use the SQL Window function to calculate adoption scores for each user 

In [14]:
# run the SQL queries. The result of each query is stored in a dataframe.
with engine.connect() as con:
    
    rs = con.execute("SELECT user_id" 
                     "       , SUM(CASE WHEN logins >= 3 THEN 1 ELSE 0 END) AS adoption_score "
                     "FROM ("
                      "      SELECT user_id, CAST(T.time_stamp AS DATE) AS dte" 
                      "                              , COUNT(*) OVER (partition by user_id "
                      "                                ORDER BY CAST(T.time_stamp AS DATE) " 
                      "                                RANGE BETWEEN INTERVAL '6' DAY PRECEDING AND CURRENT ROW) AS logins "
                      "      FROM takehome_user_engagement AS T "  
                      "     ) WindowTable "
                      "GROUP BY user_id")
    q_3 = pd.DataFrame(rs.fetchall())
    q_3.columns = rs.keys()

q_3.adoption_score = q_3.adoption_score.astype(np.int64) 
q_3

user_id  adoption_score
0           1               0
1           2               2
2           3               0
3           4               0
4           5               0
...       ...             ...
8818    11996               0
8819    11997               0
8820    11998               0
8821    11999               0
8822    12000               0

[8823 rows x 2 columns]

For example in the above user_id "2" has an adoption score of "2" because there are two 7-day windows when the user is considered "adopted" or in other words 2 windows when the user has logged-on on three separate days. Keep in mind that these windows may partially overlap. Also interesting is that this table ends up having 8823 rows where each row corresponds to one user. If you scroll above you can also see that 'last_session_creation_time' column in user_df dataframe also has 8823 non-null value rows which means that the 12,000-8,823=3,177 users which don't have a 'last_session_creation_time' recorded for them are most likely the same users that don't have any log-in activity recorded. Let's confirm this hypothesis.

In [15]:
q_3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8823 entries, 0 to 8822
Data columns (total 2 columns):
user_id           8823 non-null int64
adoption_score    8823 non-null int64
dtypes: int64(2)
memory usage: 138.0 KB


In [16]:
q_3.describe()

user_id  adoption_score
count   8823.000000     8823.000000
mean    6019.821716       18.193585
std     3464.251001       67.448943
min        1.000000        0.000000
25%     3017.500000        0.000000
50%     6034.000000        0.000000
75%     9029.500000        0.000000
max    12000.000000      593.000000

In [17]:
a1 = q_3.user_id.values
a2 = users_df[users_df.last_session_creation_time.notnull()].index.values

In [18]:
np.array_equal(a1,a2)

True

The above confirms our hypothesis.

In [19]:
q_3.set_index('user_id', inplace=True)
q_3

adoption_score
user_id                
1                     0
2                     2
3                     0
4                     0
5                     0
...                 ...
11996                 0
11997                 0
11998                 0
11999                 0
12000                 0

[8823 rows x 1 columns]

Now I will add the adoption score from the log-in data stored in q_3 dataframe to the users_df dataframe by joining the two dataframes on their User IDs. After the join replace the null value of users without any adoption score with 0. 

'last_session_creation_time' and 'creation_time' columns are not in datetime format and can be converted to a datetime data type.

Also in order to deal with the null values in 'last_session_creation_time' we can replace them with the dates in the 'creation_time' column since we know that based on the confirmed hypothesis explained previously that these users never logged-on.

In [20]:
users_df.index.name = 'user_id'

In [21]:
scored_users_df = users_df.join(q_3)

In [22]:
scored_users_df.creation_time = pd.to_datetime(scored_users_df.creation_time)
scored_users_df.last_session_creation_time = pd.to_datetime(scored_users_df.last_session_creation_time, unit='s')
scored_users_df['adoption_score'] = scored_users_df.adoption_score.fillna(0)
scored_users_df.last_session_creation_time = scored_users_df.last_session_creation_time.fillna(scored_users_df.creation_time)

We can also extract the day, month and year for the date columns as separate features.

In [23]:
scored_users_df['cr_year'] = scored_users_df.creation_time.dt.year
scored_users_df['cr_month'] = scored_users_df.creation_time.dt.month
scored_users_df['cr_day'] = scored_users_df.creation_time.dt.day
scored_users_df['last_cr_year'] = scored_users_df.last_session_creation_time.dt.year
scored_users_df['last_cr_month'] = scored_users_df.last_session_creation_time.dt.month
scored_users_df['last_cr_day'] = scored_users_df.last_session_creation_time.dt.day

In order to be able to create a binary classification model I will define a "adopted" column which is True if the "adoption score" is more than 0 and Fals otherwise.

In [24]:
scored_users_df['adopted'] = scored_users_df.adoption_score > 0

Let us define that when a user is not invited by any another user we record the user being invited by user_id 0 since user IDs start from 1. Therefore we can reaplce the null values in invited_by_user with 0. 'invited_by_user_id', 'creation_source' and 'org_id' are categorical data columns.

We can also convert adoption score from float to int and convert both 'opted_in_to_mailing_list' and 'enabled_for_marketing_drip' to boolean data type.

In [25]:
scored_users_df.invited_by_user_id = scored_users_df.invited_by_user_id.fillna(0)
scored_users_df.invited_by_user_id = scored_users_df.invited_by_user_id.astype(np.int64)
scored_users_df.invited_by_user_id = scored_users_df.invited_by_user_id.astype('category')
scored_users_df.creation_source = scored_users_df.creation_source.astype('category')
scored_users_df.org_id = scored_users_df.org_id.astype('category')
scored_users_df.adoption_score = scored_users_df.adoption_score.astype(np.int64)
scored_users_df.opted_in_to_mailing_list = scored_users_df.opted_in_to_mailing_list.astype(bool)
scored_users_df.enabled_for_marketing_drip = scored_users_df.enabled_for_marketing_drip.astype(bool)

In [26]:
scored_users_df.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 12000 entries, 1 to 12000
Data columns (total 17 columns):
creation_time                 12000 non-null datetime64[ns]
name                          12000 non-null object
email                         12000 non-null object
creation_source               12000 non-null category
last_session_creation_time    12000 non-null datetime64[ns]
opted_in_to_mailing_list      12000 non-null bool
enabled_for_marketing_drip    12000 non-null bool
org_id                        12000 non-null category
invited_by_user_id            12000 non-null category
adoption_score                12000 non-null int64
cr_year                       12000 non-null int64
cr_month                      12000 non-null int64
cr_day                        12000 non-null int64
last_cr_year                  12000 non-null int64
last_cr_month                 12000 non-null int64
last_cr_day                   12000 non-null int64
adopted                       12000 non-null boo

In [27]:
scored_users_df.adoption_score.describe()

count    12000.000000
mean        13.376833
std         58.388824
min          0.000000
25%          0.000000
50%          0.000000
75%          0.000000
max        593.000000
Name: adoption_score, dtype: float64

In [28]:
scored_users_df.adoption_score.quantile([.80, .85, .90, .95, 0.99])

0.80      0.0
0.85      0.0
0.90      6.0
0.95     80.0
0.99    343.0
Name: adoption_score, dtype: float64

As can be seen the dataframe is very unbalanced when looking at the adoption_score column. 85% of the dataframe consists of a 0 score so a model always predicting 0 would likely yeild 85% accuracy. Therefore it may be a good idea to oversample rows having an adoption score greater than 0.

Let us try to build a model which predicts the "adoption_score" of a user and check which features played a more significant role in the model. Since we are trying to predict a score number this is a regression problem.

In [29]:
scored_users_df['adopted_num'] = scored_users_df['adopted'].astype(int) 

Some features which can be dropped are 'name' and 'email' as it is unlikely that these featues would be good predictors of the user adoption score.

In [30]:
X = scored_users_df.drop(['name','email','adoption_score','creation_time','last_session_creation_time','adopted'], axis=1)
y_regression = scored_users_df[['adoption_score']]
y_classification = scored_users_df[['adopted']]

In [37]:
X[X.adopted_num == 0].groupby(['last_cr_year','last_cr_month']).count()['adopted_num']

last_cr_year  last_cr_month
2012          5                   5
              6                 314
              7                 320
              8                 292
              9                 356
              10                333
              11                332
              12                347
2013          1                 383
              2                 333
              3                 396
              4                 383
              5                 385
              6                 382
              7                 405
              8                 417
              9                 403
              10                439
              11                450
              12                431
2014          1                 482
              2                 454
              3                 576
              4                 635
              5                1076
              6                  69
Name: adopted_num, dtype: int64

In [36]:
X[X.adopted_num == 1].groupby(['last_cr_year','last_cr_month']).count()['adopted_num']

last_cr_year  last_cr_month
2012          7                  3
              8                  3
              9                  9
              10                14
              11                11
              12                11
2013          1                 15
              2                 16
              3                  8
              4                 21
              5                 20
              6                 18
              7                 21
              8                 14
              9                 22
              10                17
              11                14
              12                23
2014          1                 19
              2                 19
              3                 25
              4                 29
              5                611
              6                639
Name: adopted_num, dtype: int64

I guess the above means that if a user has logged on three separate days in in the past months then it is likely that the user has adopted. 

In [35]:
X

creation_source  opted_in_to_mailing_list  \
user_id                                                 
1              GUEST_INVITE                      True   
2                ORG_INVITE                     False   
3                ORG_INVITE                     False   
4              GUEST_INVITE                     False   
5              GUEST_INVITE                     False   
...                     ...                       ...   
11996            ORG_INVITE                     False   
11997    SIGNUP_GOOGLE_AUTH                     False   
11998          GUEST_INVITE                      True   
11999     PERSONAL_PROJECTS                     False   
12000                SIGNUP                     False   

         enabled_for_marketing_drip org_id invited_by_user_id  cr_year  \
user_id                                                                  
1                             False     11              10803     2014   
2                             False      1                316     2013   
3                             False     94               1525     2013   
4                             False      1               5151     2013   
5                             False    193               5240     2013   
...                             ...    ...                ...      ...   
11996                         False     89               8263     2013   
11997                         False    200                  0     2013   
11998                          True     83               8074     2014   
11999                         False      6                  0     2012   
12000                          True      0                  0     2014   

         cr_month  cr_day  last_cr_year  last_cr_month  last_cr_day  \
user_id                                                               
1               4      22          2014              4           22   
2              11      15          2014              3           31   
3               3      19          2013              3           19   
4               5      21          2013              5           22   
5               1      17          2013              1           22   
...           ...     ...           ...            ...          ...   
11996           9       6          2013              9            6   
11997           1      10          2013              1           15   
11998           4      27          2014              4           27   
11999           5      31          2012              6            2   
12000           1      26          2014              1           26   

         adopted_num  
user_id               
1                  0  
2                  1  
3                  0  
4                  0  
5                  0  
...              ...  
11996              0  
11997              0  
11998              0  
11999              0  
12000              0  

[12000 rows x 12 columns]

In [ ]:
y_classification

For many of the models to work we need to make a one-hot-encoding for 'creation_source'

In [ ]:
# X.creation_time = X.creation_time.apply(lambda x: x.toordinal())
# X.last_session_creation_time = X.last_session_creation_time.apply(lambda x: x.toordinal())
X = pd.get_dummies(X, columns=['creation_source'])

In [ ]:
pd.concat([X,y_classification], axis=1)

In [ ]:
pd.concat([X,y_regression], axis=1)

In [ ]:
plt.matshow(pd.concat([X,y_regression], axis=1).corr())
plt.show()
pd.concat([X,y_regression], axis=1).corr()

In [ ]:
plt.matshow(pd.concat([X,y_classification], axis=1).corr())
plt.show()
pd.concat([X,y_classification], axis=1).corr()

The only column appearing to be corelated to the class label is 'last_cr_year' which is the user's last log-in year.

In [ ]:
# from imblearn.over_sampling import RandomOverSampler

In [ ]:
# sm = RandomOverSampler(sampling_strategy='not majority', random_state=42)

In [ ]:
# X_res, y_res = sm.fit_resample(X, y.values.ravel())

In [ ]:
from sklearn.model_selection import train_test_split
X_train_class, X_test_class, y_train_class, y_test_class = \
                       train_test_split(X, y_classification, test_size=0.20, random_state=42)

In [ ]:
X_train_regres, X_test_regres, y_train_regres, y_test_regres = \
                       train_test_split(X, y_regression, test_size=0.20, random_state=42)

In [ ]:
from sklearn.ensemble import RandomForestClassifier

In [ ]:
clf = RandomForestClassifier(n_estimators=100, max_depth=5, random_state=0)

In [ ]:
clf.fit(X_train_class, y_train_class.values.ravel())

Let us first try with a Random Forest Regressor.

In [ ]:
# feature importance
print(clf.feature_importances_)
# plot
plt.bar(range(len(clf.feature_importances_)), clf.feature_importances_)
plt.show()

In [ ]:
X_train_class

Based on the above it appears that 'last_cr_year' , and 'last_cr_month' are best predictors of the user being adopted or not. These are the year and month of the user's last log-in. After that it is 'cr_year' and 'cr_month'. These are the year and month of the user being created.

In [ ]:
print(clf.score(X_test_class, y_test_class))
print(clf.score(X_train_class, y_train_class))
y_test_class_pred = clf.predict(X_test_class)

In [ ]:
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report, roc_curve, auc

In [ ]:
print(confusion_matrix(y_test_class_pred, y_test_class))
print(accuracy_score(y_test_class_pred, y_test_class))
print(classification_report(y_test_class_pred, y_test_class))
fpr, tpr, thresholds = roc_curve(y_test_class_pred, y_test_class)
plt.figure()
plt.plot(fpr, tpr, color='darkorange', lw=2, label='ROC curve (area = %0.2f)' % auc(fpr, tpr))
plt.plot([0, 1], [0, 1], color='navy', lw=2, linestyle='--')
plt.xlim([0.0, 1.0])
plt.ylim([0.0, 1.05])
plt.xlabel('False Positive Rate')
plt.ylabel('True Positive Rate')
plt.title('Receiver operating characteristic example')
plt.legend(loc="lower right")
plt.show()

In [ ]:
from sklearn.ensemble import RandomForestRegressor

In [ ]:
regr = RandomForestRegressor(max_depth=5, random_state=0,n_estimators=100)
regr.fit(X_train_regres, y_train_regres.values.ravel())

In [ ]:
# feature importance
print(regr.feature_importances_)
# plot
plt.bar(range(len(regr.feature_importances_)), regr.feature_importances_)
plt.show()

In [ ]:
print(regr.score(X_test_regres,y_test_regres))
print(regr.score(X_train_regres,y_train_regres))

Next we can try Linear Regression.

In [ ]:
from sklearn.linear_model import LinearRegression

In [ ]:
model = LinearRegression()
model.fit(X_train_regres, y_train_regres)

In [ ]:
print(model.score(X_test_regres,y_test_regres))

In [ ]:
# The coefficients
print('Coefficients: \n', model.coef_)

Polynomial Regression

In [ ]:
from sklearn.preprocessing import PolynomialFeatures
from sklearn.metrics import mean_squared_error, r2_score

In [ ]:
poly_features = PolynomialFeatures(degree=2)

In [ ]:
# transforms the existing features to higher degree features.
X_train_poly = poly_features.fit_transform(X_train_regres)

In [ ]:
# fit the transformed features to Linear Regression
poly_model = LinearRegression()
poly_model.fit(X_train_poly, y_train_regres)

In [ ]:
# predicting on training data-set
y_train_predicted = poly_model.predict(X_train_poly)

In [ ]:
# predicting on test data-set
y_test_predict = poly_model.predict(poly_features.fit_transform(X_test_regres))

In [ ]:
# evaluating the model on training dataset
rmse_train = np.sqrt(mean_squared_error(y_train_regres, y_train_predicted))
r2_train = r2_score(y_train_regres, y_train_predicted)
  
# evaluating the model on test dataset
rmse_test = np.sqrt(mean_squared_error(y_test_regres, y_test_predict))
r2_test = r2_score(y_test_regres, y_test_predict)
  
print("The model performance for the training set")
print("-------------------------------------------")
print("RMSE of training set is {}".format(rmse_train))
print("R2 score of training set is {}".format(r2_train))
  
print("\n")
  
print("The model performance for the test set")
print("-------------------------------------------")
print("RMSE of test set is {}".format(rmse_test))
print("R2 score of test set is {}".format(r2_test))

In [ ]:
# The coefficients
print('Coefficients: \n', poly_model.coef_)

Bayesian Ridge Regression

In [ ]:
from sklearn import linear_model
reg = linear_model.BayesianRidge()
reg.fit(X_train_regres, y_train_regres.values.ravel())

In [ ]:
print(reg.score(X_test_regres,y_test_regres))
print(reg.score(X_train_regres,y_train_regres))